zebra-network/src/peer/connection.rs

```diff
@@ -506,123 +509,143 @@ where
+                        },
+                        tx,
+                        span,
+                    }),
+                    Err(e) => Err((e, Some(tx))),
+                }
+            }
+            (AwaitingRequest, FindBlocks { known_blocks, stop }) => {
+                match self
+                    .peer_tx
+                    .send(Message::GetBlocks { known_blocks, stop })
+                    .await
+                {
+                    Ok(()) => Ok(AwaitingResponse {
+                        handler: Handler::FindBlocks,
+                        tx,
+                        span,
+                    }),
+                    Err(e) => Err((e, Some(tx))),
+                }
+            }
+            (AwaitingRequest, FindHeaders { known_blocks, stop }) => {
+                match self
+                    .peer_tx
+                    .send(Message::GetHeaders { known_blocks, stop })
+                    .await
+                {
+                    Ok(()) => Ok(AwaitingResponse {
+                        handler: Handler::FindHeaders,
+                        tx,
+                        span,
+                    }),
+                    Err(e) => Err((e, Some(tx))),
+                }
+            }
+            (AwaitingRequest, MempoolTransactions) => {
+                match self.peer_tx.send(Message::Mempool).await {
+                    Ok(()) => Ok(AwaitingResponse {
+                        handler: Handler::MempoolTransactions,
+                        tx,
+                        span,
+                    }),
+                    Err(e) => Err((e, Some(tx))),
+                }
+            }
             (AwaitingRequest, PushTransaction(transaction)) => {
                 // Since we're not waiting for further messages, we need to
                 // send a response before dropping tx.
                 let _ = tx.send(Ok(Response::Nil));
-                self.peer_tx
-                    .send(Message::Tx(transaction))
-                    .await
-                    .map_err(|e| e.into())
-                    .map(|()| AwaitingRequest)
+                match self.peer_tx.send(Message::Tx(transaction)).await {
+                    Ok(()) => Ok(AwaitingRequest),
+                    Err(e) => Err((e, None)),
+                }
             }
             (AwaitingRequest, AdvertiseTransactions(hashes)) => {
                 let _ = tx.send(Ok(Response::Nil));
-                self.peer_tx
+                match self
+                    .peer_tx
                     .send(Message::Inv(hashes.iter().map(|h| (*h).into()).collect()))
                     .await
-                    .map_err(|e| e.into())
-                    .map(|()| AwaitingRequest)
+                {
+                    Ok(()) => Ok(AwaitingRequest),
+                    Err(e) => Err((e, None)),
+                }
             }
             (AwaitingRequest, AdvertiseBlock(hash)) => {
                 let _ = tx.send(Ok(Response::Nil));
```

Comment:
Move these sends to the outer match

Reply:
Fixed in 97137259ad35d6ecaa5231794373fb9a47e3d702.

This change also fixes a bug where `AwaitingRequest` would not send errors back to the caller.

Reply:
It's not as clean as I might like.

I tried to just move `tx` into the `AwaitingResponse`s, but rustc couldn't prove that tx was being moved correctly, so I got compiler errors.